packages/runtime/src/runtime/hooks/useEnsembleAction.tsx

```diff
@@ -120,6 +131,189 @@ export const useInvokeApi: EnsembleActionHook<InvokeAPIAction> = (action) => {
+    };
+  }, [inputEl]);
+
+  const callback = useCallback((): void => {
+    try {
+      inputEl.click();
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.error(error);
+    }
+  }, [inputEl]);
+
+  return { callback };
+};
+
+export const useUploadFiles: EnsembleActionHook<UploadFilesAction> = (
+  action?: UploadFilesAction,
+) => {
+  const [body, setBody] = useState<Record<string, unknown>>();
+  const [headers, setHeaders] = useState<Record<string, unknown>>();
+  const [status, setStatus] = useState<UploadStatus>("pending");
+  const [progress, setProgress] = useState<number>(0.0);
+
+  const onCompleteAction = useEnsembleAction(action?.onComplete);
+  const onErrorAction = useEnsembleAction(action?.onError);
+  const screenContext = useScreenContext();
+
+  useRegisterBindings(
+    {
+      body,
+      headers,
+      status,
+      progress,
+    },
+    action?.id,
+    {
+      setBody,
+      setHeaders,
+      setStatus,
+      setProgress,
+    },
+  );
+
+  const callback = useCallback(async (): Promise<void> => {
+    const apiModel = screenContext?.model?.apis?.find(
+      (model) => model?.name === action?.uploadApi,
+    );
+    if (!apiModel) return;
+
+    const progressCallback = (progressEvent: ProgressEvent): void => {
+      const percentCompleted =
+        (progressEvent.loaded * 100) / progressEvent.total;
+
+      setProgress(percentCompleted);
+    };
+
+    const files = evaluate(
+      screenContext as ScreenContextDefinition,
+      action?.files,
+    ) as FileList | undefined;
+    if (!files || files.length === 0) throw Error("Files not found");
+
+    const formData = new FormData();
+    if (files.length === 1)
+      formData.append(action?.fieldName ?? "files", head(files) as Blob);
+    else
+      for (let i = 0; i < files.length; i++) {
+        formData.append(action?.fieldName ?? `file${i}`, files[i]);
+      }
+
+    const apiModelBody = apiModel?.body ?? {};
+    for (const key in apiModelBody) {
+      const evaluatedValue = isExpression(apiModelBody[key])
+        ? evaluate(
+            screenContext as ScreenContextDefinition,
+            apiModelBody[key] as string,
+          ) || action?.inputs?.[key]
+        : apiModelBody[key];
+      formData.append(key, evaluatedValue as string);
+    }
+
+    let apiUrl = apiModel.uri;
+    apiUrl = apiUrl.replace(
+      /\${(.*?)}/g,
+      (match, p1) =>
+        (evaluate(screenContext as ScreenContextDefinition, match) as string) ||
+        (action?.inputs?.[p1] as string),
+    );
+
+    // Regular expression for text matches inside ${}
+    const matches = [...apiUrl.matchAll(/\${(.*?)}/g)];
+
+    // Replace matches with the evaluated value
+    matches.forEach((match) => {
+      apiUrl = apiUrl.replace(
+        match[0],
+        (evaluate(
+          screenContext as ScreenContextDefinition,
+          match[0],
+        ) as string) || (action?.inputs?.[match[1]] as string),
+      );
+    });
```

Comment:
Have you tried just passing a template string to evaluate? We shouldn't need to evaluate each expression separately

Reply:
Also pass in inputs as context